orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1895,5 +1895,71 @@ IMPLEMENT_GRADIENT_BUILDER(GetFakeQuantGradient) {
   return {NodeDef(OpDef{"FakeQuantGrad", kMSDomain, 1}, {GO(0), O(1)}, {GI(0)})};
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetLSTMGradient) {
+  std ::vector<ArgDef> input_args;
+
+  // Add inputs of the LSTMTraining node as inputs of the LSTMGrad node
+  // Add inputs from the source node for X, W, R, B, SL, H0, C0, P
+  // Add empty argdef for non existing inputs
+  const int input_size = GetSrcNodeInputSize();
+  for (int i = 0; i < 8; i++) {
+    if (i < input_size && I(i).Exists()) {
+      input_args.push_back(I(i));
+    } else {
+      input_args.push_back(ArgDef());
+    }
+  }
+
+  if (O(0).Exists()) {
+    input_args.push_back(O(0));  // all hidden states output of the LSTMTraining node
+  } else {
+    input_args.push_back(ArgDef());
+  }
+
+  if (O(3).Exists()) {
+    input_args.push_back(O(3));  // all cell states output of the LSTMTraining node
+  } else {
+    input_args.push_back(ArgDef());
+  }
+
+  if (O(4).Exists()) {
+    input_args.push_back(O(4));  // i, o, f, c gate computations output of the LSTMTraining node
+  } else {
+    input_args.push_back(ArgDef());
+  }
+
+  // Add gradients of the outputs of the LSTMTraining node as inputs to the LSTMGrad node
+  // Gradients of the outputs of the LSTMTraining node include grad_HAll, grad_HFinal, grad_CFinal
+  for (int o = 0; o < 3; ++o) {
+    if (GO(o).Exists() && IsGradientAvailableForSrcNodeOutput(o)) {
+      input_args.push_back(GO(o));
+    } else {
+      input_args.push_back(ArgDef());
+    }
+  }
+
+  // Add gradients of the LSTMTraining inputs as outputs of the LSTMGrad node
+  // Outputs are gradients of:
+  //   1) X (input tensor)
+  //   2) W (weight tensor)
+  //   3) R (recurrence weight tensor)
+  //   4) B (bias tensor)
+  //   5) H0 (initial hidden state tensor)
+  //   6) C0 (initial cell state tensor)
```

Comment:
sorry to be picky here, when I check the doc, looks initial hidden/cell state is non differentiable, do you think we need return the gradient for them? 

sequence_lens (optional, non-differentiable) : T1
Optional tensor specifying lengths of the sequences in a batch. If not specified - assumed all sequences in the batch to have length `seq_length`. It has shape `[batch_size]`.
initial_h (optional, non-differentiable) : T
Optional initial value of the hidden. If not specified - assumed to be 0. It has shape `[num_directions, batch_size, hidden_size]`.
initial_c (optional, non-differentiable) : T
Optional initial value of the cell. If not specified - assumed to be 0. It has shape `[num_directions, batch_size, hidden_size]`.

Reply:
I believe this is probably an error. In my opinion, we should be able to differentiate both of them. 
In many cases, the initial h and initial c could be frozen, but it is possible that they are an output from another trainable subgraph or that multiple LSTM nodes are stacked on top of each other. In which case, we should probably be able to backpropagate through the network through initial h and initial c.
![image](https://user-images.githubusercontent.com/12852605/225962007-2ad19371-cfe9-4a66-b1a5-2e3dc4384db6.png)